token/wallet.tact

```diff
@@ -2,164 +2,120 @@
+
+        Jetton.init(
+            msg.destination,
+            self.master,
+            self.wallet_code,
+            code: self.wallet_code
+        ).receive_tokens(InternalTransfer {
             query_id: msg.query_id,
             amount: msg.amount,
-            from: state.owner,
+            from: self.owner,
             response_address: msg.response_destination,
             forward_ton_amount: msg.forward_ton_amount,
             forward_payload: msg.forward_payload,
         })
-        }.serialize())
-        .send_to(to_wallet_address);
-}
-
-fn receive_tokens(
-    @mut state: WalletState, 
-    msg: InternalTransfer, 
-    sender: ContractAddress, 
-    my_ton_balance: Coins, 
-    fwd_fee: Coins, 
-    msg_value: Coins
-) {
-    verify(
-        sender === state.master || 
-            sender == WalletAddress.from_data(msg.from, state.master, state.wallet_code), 
-        "sender must be master or user wallet"
-    );
-
-    let ton_balance_before_msg = my_ton_balance - msg_value;
-    let storage_fee = STORAGE_MIN_CONSUMPTION - ton_balance_before_msg.min(STORAGE_MIN_CONSUMPTION);
-    let @mut msg_value = msg_value - (storage_fee + GAS_CONSUMPTION);
-
-    if (forward_ton_amount != 0) {
-        msg_value -= (forward_ton_amount + fwd_fee);
-        let payload = msg.forward_payload;
-        RequestBuilder.new()
-            .ihr_disabled()
-            .pay_transfer_fee_separately()
-            .money(forward_ton_amount)
-            .body(TransferNotification {
-                query_id: msg.query_id,
-                amount: msg.amount,
-                from: msg.from,
-                payload: payload,
-            })
-            .send_to(owner_address);
+        .send(
+            ihr_disabled: true,
+            can_be_bounced: true,
+            carry_message_input: true,
+        );
     }
 
-    if (!response_address.is_null() && msg_value > 0) {
-        RequestBuilder.new()
-            .ihr_disabled()
-            .ignore_action_phase_errors()
-            .money(msg_value)
-            .body(Excesses {
-                query_id: msg.query_id,
-            })
-            .send_to(response_address);
+    msg receive_tokens(prefix: 0x178d4519, transfer: InternalTransfer) {
+        verify(
+            msg.sender === self.master || 
+                sender == WalletAddress.from_data(msg.from, self.master, self.wallet_code), 
+            "sender must be master or user wallet"
+        );
+
+        let ton_balance_before_msg = self.my_balance() - msg_value;
+        let storage_fee = STORAGE_MIN_CONSUMPTION - ton_balance_before_msg.min(STORAGE_MIN_CONSUMPTION);
+        let mut msg_value = msg_value - (storage_fee + GAS_CONSUMPTION);
+
+        if (transfer.forward_ton_amount != 0) {
+            msg_value -= (transfer.forward_ton_amount + fwd_fee);
+            let payload = msg.forward_payload;
+            self.owner
+                .receive_notification(TransferNotification {
```

Comment:
(sidenote) We should probably revise naming guidelines for interfaces. Current Jetton spec uses "receive_notification", but it's a bit odd because the sender sends and receiver receives notification.

Reply:
Concretely speaking both options do not sound quite right:

```
addr.receive_notification(...).send(...)
addr.notification(...).send(...)
```

Maybe try a dedicated syntax to make sending more symmetric with the interface definition? 

Also, since we declare the message type as params and place aux attributes in the tail, why not do the same here? 

```
addr->notification(query_id:..., amount:..., ..., coins: ...)
```

(Maybe we should reserve name "coins" to match with Coins type and leave generic terms like "amount" and "value" to the app developers. Alternatively we could resort to using a sigil for all non-body attributes - @coins / $coins / etc. if we run into problems with name clashes.)

The only feature left to deal with is _send mode_ that (afaik) lives outside of the message description and invisible on the recipient's side. Most of the mode is about choosing an amount of coins to send, but there is also &2 flag for error avoidance. We need to check how exactly those modes work, which ones are used in practice and maybe split that task by different use-cases, cover common cases with clear syntax and use catch-all "send_raw_msg()" API for all the remaining (rare) situations.

Reply:
(`->` is clashing with fn return types)

How about "async/await"-style prefix familiar to Swift/Rust/JS/C# users:

```
send addr.notification(..., coins:...)
```

Send modes then could be specified like that:

```
send(total_balance, destroy, ignore_errors) add.notification(query_id:..., coins: ...)
```

Reply:
PS. Exact sendmode interpretation is here: https://github.com/ton-blockchain/ton/blob/master/crypto/block/transaction.cpp#L1451

Reply:
Updated in [d6ca35f](https://github.com/tact-lang/tact-bikeshed/pull/5/commits/d6ca35fb23d43bb86e10c754a7f34e1e3668d01f)